kernel/src/hil/kv_store.rs

```diff
@@ -0,0 +1,81 @@
+//! Interface for KV Stores
+
+use crate::returncode::ReturnCode;
+
+#[derive(Clone, Copy, PartialEq)]
+pub enum Operation {
+    None,
+    Init,
+    GetKey,
+    AppendKey,
+    InvalidateKey,
+    GarbageCollect,
+}
+
+/// Implement this trait and use `set_client()` in order to receive callbacks.
+pub trait Client<'a> {
+    /// This callback is called when any operation is complete.
+    ///
+    /// 'operation': The operation that completed
+    /// 'result': Nothing on success, 'ReturnCode' on error
+    /// 'key': A key buffer is one exists
+    /// 'value': A value buffer is one exists
+    /// 'buf': A buf buffer is one exists
+    fn operation_complete(
+        &'a self,
+        operation: Operation,
+        result: Result<(), ReturnCode>,
+        key: Option<&'static mut [u8]>,
+        value: Option<&'static mut [u8]>,
+        buf: Option<&'static mut [u8]>,
+    );
+}
+
+pub trait KVStore {
+    /// Appends the key/value pair.
+    ///
+    /// `key`: A unhashed key. This will be hashed internally. This key
+    ///        will be used in future to retrieve or remove the `value`.
```

Comment:
This makes some assumptions about hashing the key, whereas I can imagine this would normally be internal to the KV store? If there is a collision, an appropriate ReturnCode variant will be returned.

Reply:
I have changes this to a hashed key.

The reason to pass the KV store an unhashed key is to allow the store to internally hash it. Now the capsule will have to perform the hash. This has the disadvantage that the implementation doesn't control the hash, but hashing internally in the KV store ends up being tricky to get right.